File: app/src/main/java/com/github/wyang/basiclibrarydemo/MainActivity.java
package com.github.wyang.basiclibrarydemo;

import android.content.Intent;
import android.graphics.Rect;
import android.support.annotation.NonNull;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.BaseViewHolder;
import com.github.android.common.utils.ViewUtil;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

public class MainActivity extends BaseActivity {
    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.mRecyclerView)
    RecyclerView mRecyclerView;
    private Adapter mAdapter;
    private List<Item> mData;

    @Override
    public int getLayoutId() {
        return R.layout.activity_main;
    }

    @Override
    protected void initView() {
        initToolBar(mToolbar, "Demo列表", false);

        //init RecyclerView
        mRecyclerView.setLayoutManager(new GridLayoutManager(this, 2));
        mRecyclerView.addItemDecoration(new ItemDecoration(2, ViewUtil.dp2px(this, 10), true));

        //init 适配器
        mAdapter = new Adapter();
        mAdapter.setOnItemClickListener((adapter, view, position) -> {
            Item item = mData.get(position);
            if (item.activity != null)
                startActivity(new Intent(this, item.activity));
        });

        //bind RecyclerView和适配器
        mAdapter.bindToRecyclerView(mRecyclerView);

        //设置data
        mData = new ArrayList<>();
        mData.add(new Item("ComplexView", "结合google md材质设计，自定义Button", ComplexViewActivity.class));
        mData.add(new Item("底部导航", "结合google md材质设计，自定义Button", BottomNavigationActivity.class));
        mData.add(new Item("Ripple波纹动画", "xml及代码写法", RippleActivity.class));
        mData.add(new Item("自定义View", "", TestActivity.class));
        mData.add(new Item("测试用例4", "", null));
        mData.add(new Item("测试用例4", "", null));
        mData.add(new Item("测试用例4", "", null));
        mData.add(new Item("测试用例4", "", null));
        mData.add(new Item("测试用例4", "", null));
        mData.add(new Item("测试用例4", "", null));
        mData.add(new Item("测试用例4", "", null));
        mAdapter.setNewData(mData);
    }
}

class Adapter extends BaseQuickAdapter<Item, BaseViewHolder> {

    Adapter() {
        super(R.layout.item_main_recycler);
    }

    @Override
    protected void convert(BaseViewHolder helper, Item item) {
        helper.setText(R.id.tv_title, item.title);
        helper.setText(R.id.tv_description, item.description);
    }
}

class Item {
    String title;
    String description;
    Class activity;

    Item(String title, String description, Class activity) {
        this.title = title;
        this.description = description;
        this.activity = activity;
    }
}

class ItemDecoration extends RecyclerView.ItemDecoration {
    private int spanCount;
    private int spacing;
    private boolean includeEdge;

    ItemDecoration(int spanCount, int spacing, boolean includeEdge) {
        this.spanCount = spanCount;
        this.spacing = spacing;
        this.includeEdge = includeEdge;
    }

    @Override
    public void getItemOffsets(@NonNull Rect outRect, @NonNull View view,
                               @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        int position = parent.getChildAdapterPosition(view);
        int column = position % spanCount;
        if (includeEdge) {
            outRect.left = spacing - column * spacing / spanCount;
            outRect.right = (column + 1) * spacing / spanCount;
            if (position < spanCount) {
                outRect.top = spacing;
            }
            outRect.bottom = spacing;
        } else {
            outRect.left = column * spacing / spanCount;
            outRect.right = spacing - (column + 1) * spacing / spanCount;
            if (position >= spanCount) {
                outRect.top = spacing;
            }
        }
    }
}

File: app/src/main/java/com/github/wyang/basiclibrarydemo/MainActivity.kt
package com.github.wyang.basiclibrarydemo

import android.content.Intent
import android.graphics.Rect
import android.view.View
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.viewholder.BaseViewHolder
import com.github.android.common.utils.ViewUtil
import com.github.wyang.basiclibrarydemo.databinding.ActivityMainBinding
import com.github.wyang.basiclibrarydemo.databinding.LayoutToolbarBinding

class MainActivity : BaseActivity<ActivityMainBinding>() {
    private var toolbarBinding: LayoutToolbarBinding? = null
    private val mAdapter: Adapter = Adapter()
    private var mData: MutableList<Item> = ArrayList()


    override fun getViewBinding(): ActivityMainBinding {
        val binding = ActivityMainBinding.inflate(layoutInflater)
        toolbarBinding = binding.includeToolbar
        return binding
    }

    override fun initialize() {
        toolbarBinding?.let { initToolBar(it.toolbar, "控件Demo列表", false) }

        //init RecyclerView
        mBinding?.mRecyclerView?.layoutManager = GridLayoutManager(this, 2)
        mBinding?.mRecyclerView?.addItemDecoration(ItemDecoration(2, ViewUtil.dp2px(this, 10f), true))
        mBinding?.mRecyclerView?.adapter = mAdapter

        //init 适配器
        mAdapter.setOnItemClickListener { _: BaseQuickAdapter<*, *>?, _: View?, position: Int ->
            val item: Item = mData[position]
            if (item.activity != null) startActivity(Intent(this, item.activity))
        }

        //设置data
        mData.add(Item("自定义Shape", "约束布局，帧布局，线性布局，相对布局的应用", ShapeActivity::class.java))
        mData.add(Item("ComplexView", "结合google md材质设计，自定义Button", ComplexViewActivity::class.java))
        mData.add(Item("ImageView拓展", "各种ImageView应用", ImageViewActivity::class.java))
        mData.add(Item("底部导航", "结合google md材质设计，自定义Button", BottomNavigationActivity::class.java))
        mData.add(Item("轮播图-RecyclerView实现", "", BannerActivity::class.java))
        mData.add(Item("测试用例4", "", null))
        mData.add(Item("测试用例4", "", null))
        mData.add(Item("测试用例4", "", null))
        mData.add(Item("测试用例4", "", null))
        mData.add(Item("测试用例4", "", null))
        mData.add(Item("test用例", "测试", TestActivity::class.java))
        mAdapter.setNewInstance(mData)
    }
}

internal class Adapter : BaseQuickAdapter<Item, BaseViewHolder>(R.layout.item_main_recycler) {
    override fun convert(holder: BaseViewHolder, item: Item) {
        holder.setText(R.id.tv_title, item.title)
        holder.setText(R.id.tv_description, item.description)
    }
}

internal class Item(var title: String, var description: String, var activity: Class<*>?)

internal class ItemDecoration(private val spanCount: Int, private val spacing: Int, private val includeEdge: Boolean) : RecyclerView.ItemDecoration() {
    override fun getItemOffsets(outRect: Rect, view: View,
                                parent: RecyclerView, state: RecyclerView.State) {
        val position = parent.getChildAdapterPosition(view)
        val column = position % spanCount
        if (includeEdge) {
            outRect.left = spacing - column * spacing / spanCount
            outRect.right = (column + 1) * spacing / spanCount
            if (position < spanCount) {
                outRect.top = spacing
            }
            outRect.bottom = spacing
        } else {
            outRect.left = column * spacing / spanCount
            outRect.right = spacing - (column + 1) * spacing / spanCount
            if (position >= spanCount) {
                outRect.top = spacing
            }
        }
    }
}
